src/vfio_device.rs

```diff
@@ -0,0 +1,965 @@
+        let fd = unsafe { ioctl_with_ptr(self, VFIO_GROUP_GET_DEVICE_FD(), path_ptr) };
+        if fd < 0 {
+            return Err(VfioError::GroupGetDeviceFD);
+        }
+
+        // Safe as fd is valid FD
+        let device = unsafe { File::from_raw_fd(fd) };
+
+        let mut dev_info = vfio_device_info {
+            argsz: mem::size_of::<vfio_device_info>() as u32,
+            flags: 0,
+            num_regions: 0,
+            num_irqs: 0,
+        };
+        // Safe as we are the owner of dev and dev_info which are valid value,
+        // and we verify the return value.
+        let ret = unsafe { ioctl_with_mut_ref(&device, VFIO_DEVICE_GET_INFO(), &mut dev_info) };
+        if ret < 0
+            || (dev_info.flags & VFIO_DEVICE_FLAGS_PCI) == 0
+            || dev_info.num_regions < VFIO_PCI_CONFIG_REGION_INDEX + 1
+            || dev_info.num_irqs < VFIO_PCI_MSIX_IRQ_INDEX + 1
+        {
+            return Err(VfioError::VfioDeviceGetInfo);
+        }
+
+        Ok(VfioDeviceInfo {
+            device,
+            flags: dev_info.flags,
+            num_regions: dev_info.num_regions,
+            num_irqs: dev_info.num_irqs,
+        })
+    }
+}
+
+impl AsRawFd for VfioGroup {
+    fn as_raw_fd(&self) -> RawFd {
+        self.group.as_raw_fd()
+    }
+}
+
+struct VfioRegion {
+    flags: u32,
+    size: u64,
+    offset: u64,
+    mmap: (u64, u64),
+}
+
+/// Information abour VFIO interrupts.
+pub struct VfioIrq {
+    /// Flags for irq.
+    pub flags: u32,
+    /// Staring index.
+    pub index: u32,
+    /// Number interrupts.
+    pub count: u32,
+}
+
+struct VfioDeviceInfo {
+    device: File,
+    flags: u32,
+    num_regions: u32,
+    num_irqs: u32,
+}
+
+impl VfioDeviceInfo {
+    fn get_irqs(&self) -> Result<HashMap<u32, VfioIrq>> {
+        let mut irqs: HashMap<u32, VfioIrq> = HashMap::new();
+
+        for index in 0..self.num_irqs {
+            let mut irq_info = vfio_irq_info {
+                argsz: mem::size_of::<vfio_irq_info>() as u32,
+                flags: 0,
+                index,
+                count: 0,
+            };
+
+            let ret = unsafe {
+                ioctl_with_mut_ref(&self.device, VFIO_DEVICE_GET_IRQ_INFO(), &mut irq_info)
+            };
+            if ret < 0 {
+                warn!("Could not get VFIO IRQ info for index {:}", index);
+                continue;
+            }
+
+            let irq = VfioIrq {
+                flags: irq_info.flags,
+                index,
+                count: irq_info.count,
+            };
+
+            debug!("IRQ #{}", index);
+            debug!("\tflag 0x{:x}", irq.flags);
+            debug!("\tindex {}", irq.index);
+            debug!("\tcount {}", irq.count);
+
+            irqs.insert(index, irq);
+        }
+
+        Ok(irqs)
+    }
+
+    fn get_regions(&self) -> Result<Vec<VfioRegion>> {
+        let mut regions: Vec<VfioRegion> = Vec::new();
+
+        for i in VFIO_PCI_BAR0_REGION_INDEX..self.num_regions {
```

Comment:
More of a nit, but can we split this in multiple functions? For example some of the following calls can be implemented on `vfio_region_info_with_cap` instead.

Reply:
Done as well.